frontend/src/main/scala/bloop/exec/IdentifiedThread.scala

```diff
@@ -0,0 +1,25 @@
+package bloop.exec
+
+import java.util.UUID
+
+/**
+ * A `Thread` that can be identified via `id`. Its `id` is stored in an `InheritableThreadLocal`
+ * variable that can be used to perform operations specific to that thread (e.g. selecting the
+ * right outputstreams to use to simulate dedicated stdout / stderr.
+ */
+abstract class IdentifiedThread private (val id: UUID) extends Thread {
+  def this() = this(UUID.randomUUID)
+  final override def run(): Unit = {
+    IdentifiedThread.id.set(id)
+    work()
+  }
+
+  /** The main loop of this `Thread`. */
+  def work(): Unit
+}
+
+object IdentifiedThread {
+
+  /** An `InheritedThreadLocal` that can be used to identify the currently running `Thread`. */
+  val id = new InheritableThreadLocal[UUID]
```

Comment:
Have you looked into reusing `scala,util.DynamicVariable`?

Reply:
Haven't tried it, it doesn't look like it would change many things.

Reply:
This is ok then :smile:. What I'm most worried is if this will indeed work in concurrent scenarios, and we have supposedly a test for it, so I'll trust it!